src/mergeplg/interpolate.py

```diff
@@ -6,274 +6,444 @@
-            da_gauge = da_gauge.copy().expand_dims("time")
-            time_dim_was_expanded = True
-        if "time" not in da_grid.dims:
-            da_grid = da_grid.copy().expand_dims("time")
-            time_dim_was_expanded = True
-
-        # Update x0 geometry for CML and gauge
-        self.update(da_cml=da_cml, da_gauge=da_gauge)
-
-        # Get ground observations and x0 geometry
-        obs, x0 = self.get_obs_x0_(da_cml=da_cml, da_gauge=da_gauge)
-
-        # Get index of not-nan obs
-        keep = np.where(~np.isnan(obs))[0]
-
-        # Return gridded data with zeros if too few observations
-        if obs[keep].size <= self.min_observations:
-            return xr.DataArray(
-                data=[np.zeros(da_grid.x_grid.shape)],
-                coords=da_grid.coords,
-                dims=da_grid.dims,
-            )
+        # Get updated interpolator object
+        self._interpolator = self._maybe_update_interpolator(
+            da_grid, da_cmls, da_gauges
+        )
+
+        # Get correct order of observations
+        observations = self._get_obs(da_cmls, da_gauges)
 
-        # Coordinates to predict
+        # Grid specific information
         coord_pred = np.hstack(
             [da_grid.y_grid.data.reshape(-1, 1), da_grid.x_grid.data.reshape(-1, 1)]
         )
 
-        # Ensure same functionality as in kriging
-        if not nnear:
-            nnear = obs[keep].size
+        # If few observations return zero grid
+        if (~np.isnan(observations)).sum() <= self.min_observations:
+            return xr.DataArray(
+                data=np.full_like(da_grid.x_grid, np.nan),
+                coords=da_grid.coords,
+                dims=da_grid.dims,
+            )
 
-        # IDW interpolator invdisttree
-        idw_interpolator = idw.Invdisttree(x0[keep])
-        interpolated = idw_interpolator(
+        # IDW interpolator
+        interpolated = self._interpolator(
             q=coord_pred,
-            z=obs[keep],
-            nnear=obs[keep].size if obs[keep].size <= nnear else nnear,
-            p=p,
-            idw_method=idw_method,
-            max_distance=max_distance,
+            z=observations,
+            nnear=observations.size if observations.size <= self.nnear else self.nnear,
+            p=self.p,
+            idw_method=self.idw_method,
+            max_distance=self.max_distance,
         ).reshape(da_grid.x_grid.shape)
 
-        da_interpolated = xr.DataArray(
-            data=[interpolated], coords=da_grid.coords, dims=da_grid.dims
-        )
-        if time_dim_was_expanded:
-            da_interpolated = da_interpolated.isel(time=0)
-        return da_interpolated
+        return xr.DataArray(data=interpolated, coords=da_grid.coords, dims=da_grid.dims)
 
 
-class InterpolateOrdinaryKriging(Base):
-    """Interpolate CML and radar using neighbourhood ordinary kriging
+class InterpolateKriging(Interpolator):
```

Comment:
this could also be renamed to `InterpolaeKrigingBase` because it cannot be used directly